<a href="https://colab.research.google.com/github/ploybongkot/DWDM21/blob/main/MiniExam.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### ข้อสอบ

จาก data ```flights.csv``` และ ```airports.csv``` ใน https://drive.google.com/drive/folders/1EFo1_uQgWkRiXjezL0T4YwteYrsqU5MV?usp=sharing
0. หารัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดและน้อยที่สุด
1. หารัฐที่มีจำนวนเครื่องบินขาออกมากที่สุดและน้อยที่สุด
2. วาด Histogram แสดงจำนวนเครื่องบินขาออกของแต่ละวันในสัปดาห์
3. สายการบินไหนมีเที่ยวบินมากที่สุดและน้อยที่สุด
4. สายการบินอะไรมีความล่าช้ามากที่สุดและน้อยที่สุด
5. ช่วงเช้าหรือช่วงบ่ายมีเครื่องบินขาเข้ามากกว่ากัน

ให้นักศึกษาทำข้อสอบ 2 ข้อจากที่แสดงข้างบนโดยกรอกรหัสนักศึกษาลงใน function ```gen_exams()``` แล้ว run เพื่อให้ได้ข้อสอบที่นักศึกษาต้องทำ 2 ข้อ (ข้อละ 3 คะแนน)
และให้นักศึกษาคิดโจทย์เองอีก 1 ข้อ (4 คะแนน)


In [1]:
import pandas as pd

In [2]:
from google.colab import drive 
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import os

In [8]:
flights = pd.read_csv('/content/drive/MyDrive/DWDM2021_DATA/Mini Exam/flights.csv')
flights

วันในเดือน  วันในสัปดาห์  ... ความล่าช้าขาออก  ความล่าช้าขาเข้า
0                19             5  ...              -3                 1
1                19             5  ...               0                -8
2                19             5  ...              -4               -15
3                19             5  ...              28                24
4                19             5  ...              -6               -11
...             ...           ...  ...             ...               ...
2702213          19             6  ...               1                -3
2702214          19             6  ...              -2                -3
2702215          19             6  ...              18                 9
2702216          19             6  ...              10                 7
2702217          19             6  ...               0               -13

[2702218 rows x 7 columns]

In [9]:
airports = pd.read_csv('/content/drive/MyDrive/DWDM2021_DATA/Mini Exam/airports.csv')
airports

รหัสสนามบิน         เมือง รัฐ                          ชื่อสนามบิน
0          10165   Adak Island  AK                                 Adak
1          10299     Anchorage  AK  Ted Stevens Anchorage International
2          10304         Aniak  AK                        Aniak Airport
3          10754        Barrow  AK      Wiley Post/Will Rogers Memorial
4          10551        Bethel  AK                       Bethel Airport
..           ...           ...  ..                                  ...
360        11233      Cheyenne  WY  Cheyenne Regional/Jerry Olson Field
361        11097          Cody  WY                 Yellowstone Regional
362        11865      Gillette  WY             Gillette Campbell County
363        12441       Jackson  WY                         Jackson Hole
364        14543  Rock Springs  WY       Rock Springs Sweetwater County

[365 rows x 4 columns]

In [10]:
def gen_exams(ID):
    import math
    a = int(ID[-1])%6
    b = int(ID[-1])%3
    print(f'รหัสนักศึกษา {ID} ทำข้อ {a} และ {b if a != b else 5-b}')

ตัวอย่างการ run

In [11]:
gen_exams('62123456-9')

รหัสนักศึกษา 62123456-9 ทำข้อ 3 และ 0


In [12]:
gen_exams('623021051-4')

รหัสนักศึกษา 623021051-4 ทำข้อ 4 และ 1


จากการ run รหัสนักศึกษาได้ทำข้อสอบข้อ 4 และ 1



In [13]:
merged_table = flights.merge(airports,how='left',left_on='รหัสสนามบินต้นทาง',right_on='รหัสสนามบิน') ## รวม 2 ตารางเข้าด้วยกัน
merged_table

วันในเดือน  ...                                 ชื่อสนามบิน
0                19  ...                  Detroit Metro Wayne County
1                19  ...                Salt Lake City International
2                19  ...                      Portland International
3                19  ...             Lambert-St. Louis International
4                19  ...  Cincinnati/Northern Kentucky International
...             ...  ...                                         ...
2702213          19  ...                       Orlando International
2702214          19  ...    Hartsfield-Jackson Atlanta International
2702215          19  ...           Albuquerque International Sunport
2702216          19  ...    Hartsfield-Jackson Atlanta International
2702217          19  ...    Hartsfield-Jackson Atlanta International

[2702218 rows x 11 columns]

ข้อ 4 สายการบินอะไรมีความล่าช้ามากที่สุดและน้อยที่สุด

In [23]:
max2_1 = merged_table.groupby('สายการบิน').max().sort_values(by=['ความล่าช้าขาออก'],ascending=False) ## หาสายการบินที่มีความล่าช้าขาออกมากที่สุด
max2_1.iloc[:1,4:5] ## ตัดเอาแค่คอลัมน์สายการบินและคอลัมน์ความล่าช้าขาออก

ความล่าช้าขาออก
สายการบิน                 
AA                    1863

สายการบิน AA มีความล่าช้าขาออกมากที่สุด คือ 1863

In [24]:
max2_2 = merged_table.groupby('สายการบิน').max().sort_values(by=['ความล่าช้าขาเข้า'],ascending=False) ## หาสายการบินที่มีความล่าช้าขาเข้าน้อยที่สุด
max2_2.iloc[:1,5:6] ## ตัดเอาแค่คอลัมน์สายการบินและคอลัมน์ความล่าช้าขาเข้า

ความล่าช้าขาเข้า
สายการบิน                  
HA                     1845

สายการบิน HA มีความล่าช้าขาออกน้อยที่สุด คือ 1845

1. หารัฐที่มีจำนวนเครื่องบินขาออกมากที่สุดและน้อยที่สุด

In [28]:
max = merged_table.groupby('รัฐ').count().sort_values(by=['รหัสสนามบินปลายทาง'],ascending=False)  ## หารัฐที่มีจำนวนเครื่องบินขาออกมากสุด 
max.iloc[:1,4:5] ## ตัดเอาแค่คอลัมน์รัฐและคอลัมน์สนามบินปลายทาง

รหัสสนามบินปลายทาง
รัฐ                    
CA               380046

รัฐที่มีจำนวนเครื่องบินขาออกมากที่สุดคือรัฐ CA คือ 380046

In [29]:
min = merged_table.groupby('รัฐ').count().sort_values(by=['รหัสสนามบินปลายทาง'],ascending=False) ## หารัฐที่มีจำนวนเครื่องบินขาออกน้อยสุด 
min.iloc[-1:,4:5] ## ตัดเอาแค่คอลัมน์รัฐและคอลัมน์สนามบินปลายทาง

รหัสสนามบินปลายทาง
รัฐ                    
AK                 7700

In [ ]:
รัฐที่มีจำนวนเครื่องบินขาออกน้อยที่สุดคือรัฐ AK คือ 7700